presto-spi/src/main/java/io/prestosql/spi/type/TimestampType.java

```diff
@@ -13,49 +13,62 @@
  */
 package io.prestosql.spi.type;
 
-import io.prestosql.spi.block.Block;
-import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.PrestoException;
+
+import static io.prestosql.spi.StandardErrorCode.NUMERIC_VALUE_OUT_OF_RANGE;
+import static java.lang.String.format;
 
 /**
  * A timestamp is stored as milliseconds from 1970-01-01T00:00:00 UTC and is to be interpreted as date-time in UTC.
  * In legacy timestamp semantics, timestamp is stored as milliseconds from 1970-01-01T00:00:00 UTC and is to be
  * interpreted in session time zone.
  */
-public final class TimestampType
-        extends AbstractLongType
+public abstract class TimestampType
+        extends AbstractType
+        implements FixedWidthType
 {
-    public static final TimestampType TIMESTAMP = new TimestampType();
+    public static final int MAX_PRECISION = 12;
 
-    private TimestampType()
-    {
-        super(new TypeSignature(StandardTypes.TIMESTAMP));
-    }
+    public static final int MAX_SHORT_PRECISION = 6;
+    private static final int DEFAULT_PRECISION = 3; // TODO: should be 6 per SQL spec
 
-    @Override
-    public Object getObjectValue(ConnectorSession session, Block block, int position)
+    public static final TimestampType TIMESTAMP = createTimestampType(DEFAULT_PRECISION);
```

Comment:
`@deprecate`, since it should not be used to determine something is a timestamp type (as in https://github.com/prestosql/presto/pull/3762)